src/internal/connector/sharepoint/collection.go

```diff
@@ -148,52 +159,59 @@ func (sc *Collection) populate(ctx context.Context) {
 
 	defer func() {
 		close(colProgress)
-		sc.finishPopulation(ctx, success, totalBytes, errs)
+		sc.finishPopulation(ctx, objects, success, totalBytes, errs)
 	}()
 
-	// sc.jobs contains query = all of the site IDs.
-	for _, id := range sc.jobs {
-		// Retrieve list data from M365
-		lists, err := loadLists(ctx, sc.service, id)
+	// Retrieve list data from M365
+	lists, err := loadSiteLists(ctx, sc.service, sc.fullPath.ResourceOwner(), sc.jobs)
+	if err != nil {
+		errs = support.WrapAndAppend(sc.fullPath.ResourceOwner(), err, errs)
+	}
+
+	objects += len(lists)
+	// Write Data and Send
+	for _, lst := range lists {
+		byteArray, err := serializeListContent(writer, lst)
 		if err != nil {
-			errs = support.WrapAndAppend(id, err, errs)
+			errs = support.WrapAndAppend(*lst.GetId(), err, errs)
+			continue
 		}
-		// Write Data and Send
-		for _, lst := range lists {
-			err = writer.WriteObjectValue("", lst)
-			if err != nil {
-				errs = support.WrapAndAppend(*lst.GetId(), err, errs)
-				continue
-			}
 
-			byteArray, err := writer.GetSerializedContent()
-			if err != nil {
-				errs = support.WrapAndAppend(*lst.GetId(), err, errs)
-				continue
+		arrayLength = int64(len(byteArray))
+
+		if arrayLength > 0 {
+			t := time.Now()
+			if t1 := lst.GetLastModifiedDateTime(); t1 != nil {
+				t = *t1
 			}
 
-			writer.Close()
+			totalBytes += arrayLength
 
-			arrayLength = int64(len(byteArray))
+			success++
+			sc.data <- &Item{
+				id:      *lst.GetId(),
+				data:    io.NopCloser(bytes.NewReader(byteArray)),
+				info:    sharePointListInfo(lst, arrayLength),
+				modTime: t,
+			}
 
-			if arrayLength > 0 {
-				t := time.Now()
-				if t1 := lst.GetLastModifiedDateTime(); t1 != nil {
-					t = *t1
-				}
+			colProgress <- struct{}{}
+		}
+	}
+}
 
-				totalBytes += arrayLength
+func serializeListContent(writer *kw.JsonSerializationWriter, lst models.Listable) ([]byte, error) {
+	defer writer.Close()
```

Comment:
This deferral doesn't protect from panics.  Since the writer (line 153) exists external to this func, and is not reconstructed on each iteration of the for loop, any len of lists > 1 will still cause a panic.  The `defer close()` needs to be co-located with the instance of the writer.  That means that either this func should generate a writer, rather than accept one; or the close should move out of this func.

If we're handling an iteration of items, we probably want the former behavior (creating the writer within this func).  I don't know that we'll get consistent or expected behavior from re-using a writer in the iteration, either.

Reply:
@ryanfkeepers We are probably disagreeing about how this function works. If a SerializationWriter is not closed, it does not cause panic. This is a helper function that I created in order to reuse the writer. This is the code [link](https://github.com/microsoft/kiota-serialization-json-go/blob/598b061421655f2d8dcc6541d4b75c0b9c1d7754/json_serialization_writer.go#L730)
```go
func (w *JsonSerializationWriter) Close() error {
	w.writer = make([]string, 0)
	return nil
}
```

For efficiency's sake, it is overkill to recreate the object in every loop.

Reply:
Mmm, fair. I disagree on the naming of `Close`, in this case, because it carries the implication that the resource will be unusable afterwards.  `Reset` would be a better func name here.

Either way, thanks for the insight.  That's enough to unblock things for me.